builder/builder.go

```diff
@@ -1151,6 +1151,55 @@ func (b *Builder) EditBundles(bundles []string, suppressEditor bool, add bool, g
 	return nil
 }
 
+// ValidateLocalBundles runs bundle parsing validation on all local bundles.
+func (b *Builder) ValidateLocalBundles(lvl ValidationLevel) error {
+	files, err := ioutil.ReadDir(b.LocalBundleDir)
+	if err != nil {
+		return errors.Wrap(err, "Failed to read local-bundles")
+	}
+
+	invalid := false
+	for _, file := range files {
+		if err := validateBundleFile(filepath.Join(b.LocalBundleDir, file.Name()), lvl); err != nil {
+			invalid = true
+			fmt.Printf("Invalid: %q:\n%s\n\n", file.Name(), err)
+		}
+	}
+
+	if invalid {
+		return errors.New("Invalid bundles found")
+	}
+
+	return nil
+}
+
+// ValidateBundles runs bundle parsing validation on a list of local bundles. In
+// addition to parsing errors, errors are generated if the bundle is not found
+// in local-bundles.
+func (b *Builder) ValidateBundles(bundles []string, lvl ValidationLevel) error {
+	invalid := false
+	for _, bundle := range bundles {
+		path := filepath.Join(b.LocalBundleDir, bundle)
+
+		if _, err := os.Stat(path); err != nil {
```

Comment:
Do you need this check here? Just a bit down the line, parseBundleFile will fail if the file doesn't exist. Also, if you remove this check, you can change ValidateLocalBundles to just call ValidateBundles passing the files list.

Reply:
You're right that `validateBundleFile` would eventually die out when `parseBundleFile` tries to open the file. But a) the error message you get back there would be a little cryptic, and b) I really wanted to make it explicit that what's wrong in this case is that the bundle _isn't found in local bundles_. Almost everywhere else, the tool will "automatically look up where the bundle is located, looking first in local bundles and then in upstream" or whatever, but that's not the case here. This checks _only_ local bundles (because it makes no sense for a user to check upstream bundles), and I wanted to make that clear.

Even as is, `ValidateLocalBundles` could just call `ValidateBundles`, but (ignoring the redundant local check you're suggesting I remove) it'd mean looping through the list of bundles a second time (once to form the `[]string` out of the `[]FileInfo` in `ValidateLocalBundles`, and then again to call `validateBundleFile` down in `ValidateBundles`). I considered doing this, but figured `validateBundleFile` was enough of an abstraction/re-use.

Reply:
I don't like the idea that if you change one function you mostly likely gonna have to remember to change the other. You can check the error type to have something less cryptic if that is the issue. This would save a stat operation and any cryptic TOCTOU error that could eventually show up.

Reply:
I checked what the error output is, and it was less cryptic than I had thought -- you basically just get a "does not exist" message:
```
clr@machine /home/clr/mix # mixer bundle validate foobar
Invalid: "foodbar":
open /home/clr/mix/local-bundles/foobar: no such file or directory
```

As a result, I pulled this out and had `ValidateLocalBundles` just call directly into this like you suggested.